File: IntegerPairs/model_solution/integer_pairs.py
'''
Integer Pairs
favoring memory efficiency over time efficiency
O(n log n) runtime due to sorting
with O(1) space
'''

# def integerPairs(arr, k):
  # found_pair = False
  # arr.sort()
  # first = 0
  # last = len(arr) - 1

  # while first < last:
    # sum = arr[first] + arr[last]

    # if sum == k:
      # print(str(arr[first]) + ', ' + str(arr[last]))
      # first += 1
      # last -= 1
      # found_pair = True
    # else:
      # if sum < k:
        # first += 1
      # else:
        # last -= 1
  # if not found_pair:
    # print('No pairs found')

'''
Integer Pairs
trades time efficiency for space efficiency
O(n) runtime with O(n) space
'''

def integerPairs(arr, k):
  found_pair = False
  hash = {}

  for elem in arr:
    if k - elem in hash:
      print(elem, k - elem)
      found_pair = True
    else:
      hash[elem] = True
  if not found_pair:
    print("No pairs found")


# Some basic tests
integerPairs([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 11)   # should print '1, 10', '2, 9', '3, 8', '4, 7', '5, 6'
print('---------------')
integerPairs([5, 5, 4], 12)                         # should print 'No pairs found'
print('---------------')
integerPairs([99, 45, 38, 1, 68, 78], 100)          # should print '1, 99'

File: IntegerPairs/model_solution/integerPairs.js
//A runtime-efficient implementation that trades time efficiency for space efficiency

//O(n) runtime with O(n) space

//function integerPairs(arr, k) {
  //const hash = {}

  //arr.forEach((x, i) => {
    //if (hash[k - x]) {
      //console.log(x, k - x)
    //} else {
      //hash[x] = i + 1
    //}
  //})
//}

//Alternative implementation that favors memory efficiency over time efficiency
//O(n log n) runtime due to the sorting with O(1) space

function integerPairs(arr, k) {
  arr.sort((x, y) => x - y)

  let first = 0
  let last = arr.length - 1

  while (first < last) {
    const sum = arr[first] + arr[last]

    if (sum == k) {
      console.log(arr[first], arr[last])
      first++
      last--
    } else {
      if (sum < k) first++;
      else last--;
    }
  }
}

/* Some console.log tests */
integerPairs([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 11);  // should print '6 5', '7 4', '8 3', '9 2', '10 1'
console.log();
integerPairs([5, 5, 4], 12);                        // should not print anything
console.log();
integerPairs([99, 45, 38, 1, 68, 78], 100);         // should print '1 99'
